### DISCUSSION TITLE
[feat][v1] How to support both beta and stable versions

### DISCUSSION THREAD

**Discussion by ayangweb** [tauri-apps/tauri#11069] [Q&A] [answered]
### Describe the problem

Hey 👋, what should I do if I want to make it so that users who have beta receiving turned on in v1 receive the beta version, and those who don't have it turned on receive only the stable version?

### Describe the solution you'd like

Get Updates: https://github.com/ayangweb/EcoPaste-Updater/blob/master/api/update.ts

update config: https://github.com/EcoPasteHub/EcoPaste/blob/ba343b2fd836fb60430e39c03d86f21e8f61b730/src-tauri/tauri.conf.json#L118

If my current stable version is v0.1.0, but some users did not update to v0.1.0 in time before the release of v0.1.1-beta.1, then they can't receive the v0.1.0 update. How should I fix this issue please? Thanks!

**Comment by amrbashir** [comment, marked as answer]
You will need to change the endpoint in your config to be something like this `"https://ecopaste-updater.ayangweb.cn/api/update/{{current_version}}"` which will be replaced by tauri updater with the current tauri app version, so your users will request updates from something like this `https://ecopaste-updater.ayangweb.cn/api/update/0.1.0` and in your updater server, you can check this version and provide the necessary update as you see fit.

**Reply by HuakunShen** [reply]
Thank you!
In my understanding, when the updater server receives a stable version like `0.1.0`, it responds with the latest stable version. Similarly, if it receives a beta version like `0.1.0-beta.0`, it returns the latest beta release. This approach, however, creates two entirely separate update channels: one for stable versions and another for beta versions.

The challenge arises when we want to give users the option to opt into beta updates, even if they’re currently on a stable version. Using the `/{{current_version}}` method, the app is restricted to updating within its respective channel—beta versions can only update to other betas, and stable versions to other stables.

**Reply by HuakunShen** [reply]
In Tauri V2, updater's `check()` method accepts `headers`. I was able to tell the update server which update channel I want like this.

```ts
const beta: boolean = false
const update = await check({
	headers: {
		"updater-mode": beta ? "beta" : "stable"
	}
})
```

But Tauri V1 doesn't have the headers option.

**Reply by amrbashir** [reply]
v1 has the headers option in the Rust API.

**Reply by amrbashir** [reply]
Using the Rust API in v1 or v2, you could also conditionally change the endpoint, so if your users is on beta you request updates from "https://ecopaste-updater.ayangweb.cn/api/update/beta/{{current_version}}" and if on stable, you request from "https://ecopaste-updater.ayangweb.cn/api/update/{{current_version}}"

https://docs.rs/tauri/latest/tauri/updater/struct.UpdateBuilder.html#method.endpoints

**Reply by HuakunShen** [reply]
Thanks! This is very helpful.

**Comment by ayangweb** [comment]
@amrbashir @HuakunShen Thanks for all your help🌹, I know how to fix the problem!🥳
